### PR TITLE
3.4.1

### PR DESCRIPTION
**Pull request by Nexela**
Far from ready for any merges. Also fully mostly annotated

### CODE REVIEW

File: factorio-plugin/require.lua [hunk 1]
```diff
@@ -2,23 +2,18 @@
...
-  ---@type string|number
-  for start, name, finish in text:gmatch("require%s*%(?%s*['\"]()(.-)()['\"]%s*%)?") do
-    ---@type string
+local function replace(_, text, diffs)
+  for start, name, finish in
+    text:gmatch("require%s*%(?%s*['\"]()(.-)()['\"]%s*%)?")--[[@as fun(): integer, string, integer]]
+  do
+
     local original_name = name
-    -- if name has slashes, convert to a dotted path, because the
```
Comment (Nexela) [inline factorio-plugin/require.lua:13]: Sumneko allows setting the path separator per workspace/folder. As long is it is correctly set in json for the type the mod uses this won't be an issue.

File: plugin.lua [hunk 2]
```diff
@@ -45,19 +45,18 @@ if not __plugin_dev and not _G.__factorio_plugin_initialized then
 
     plugin_path = fs.path(workspace.getAbsolutePath(scp.uri, config.get(scp.uri, 'Lua.runtime.plugin')))
   else -- sumneko.lua < 2.6.0
+    ---@diagnostic disable-next-line: missing-parameter
     plugin_path = fs.path(workspace.getAbsolutePath(config.get('Lua.runtime.plugin')))
   end
 
-  ---@type string
-  local new_path = (plugin_path:parent_path() / "?.lua"):string()
+  local new_path = (plugin_path:parent_path() / "?.lua"):string() --[[@as string]]
```
Comment (Nexela) [inline plugin.lua]: This `@as` probably isn't needed but I don't have bee linked in the library so I could not confirm it.

File: plugin.lua [hunk 3]
```diff
@@ -45,19 +45,18 @@ if not __plugin_dev and not _G.__factorio_plugin_initialized then
...
   if not package.path:find(new_path, 1, true) then
     package.path = package.path..";"..new_path
   end
 end
 
 local require_module = require("factorio-plugin.require")
 local global = require("factorio-plugin.global")
-local narrow = require("factorio-plugin.narrow")
```
Comment (Nexela) [inline plugin.lua:63]: Supported with `@cast` and `@as` in 3.0ish

File: plugin.lua [hunk 4]
```diff
@@ -7,19 +7,19 @@
...
 -- allow for require to search relative to this plugin file
 -- open for improvements!
+
+---Notes: confirm "path/to/lua-language-server/script/", in Lua.Workspace.Library
+---Notes: confirm "path/to/bee" in Lua.Workspace.Library
+
 if not __plugin_dev and not _G.__factorio_plugin_initialized then
   _G.__factorio_plugin_initialized = true
 
-  ---@type table
```
Comment (Nexela) [inline plugin.lua:14]: Inferred correctly when "path/to/lua-language-server/script/", is in Lua.Workspace.Library

File: plugin.lua [hunk 5]
```diff
@@ -71,13 +70,21 @@ local on_event = require("factorio-plugin.on-event")
 ---@param text string @ The content of file
 ---@return nil|Diff[]
 function OnSetText(uri, text)
-  if text:sub(1, 4) == "--##" then return end
+  ---I can't see a reason to process ---@meta files
+  ---Speeds up loading by not reading mod debugger annotations.
+  local sub = text:sub(1, 8)
+  if sub == "---@meta" or sub == "--##" then return end
```
Comment (Nexela) [inline plugin.lua]: Ignore ---@meta files?  Thoughts? It improves load times when Factorio typings are in the library

File: factorio-plugin/on-event.lua [hunk 6]
```diff
@@ -3,57 +3,57 @@
...
 ---@param diffs Diff[] @ The diffs to add more diffs to
-local function replace(uri, text, diffs)
-  ---@param s_func_param number
+local function replace(_, text, diffs)
+  ---@param s_func_param integer
   ---@param class_name_getter fun(): string
   local function process_func_param(s_func_param, class_name_getter)
-    ---@type string|number
-    local s_func, param_name, f_func = text:match("^%s*()function%s*%(%s*([^)%s]+)()", s_func_param)
+    local s_func, ---@type integer|nil
```
Comment (Nexela) [inline factorio-plugin/on-event.lua]: Its ugly, but until official type list type support is added it works.

File: plugin.lua (hunk 2 above)
Comment (Nexela) [reply plugin.lua]: Bee is now included in meta so it doesn't need library path added

File: plugin.lua (hunk 5 above)
Comment (JanSharp) [reply plugin.lua]: Pretty sure there once was something in those files that needed to be touched by the plugin but right now that is not the case. However jarg has been talking about a custom `---@event` tag which the plugin would have to see. But since that is in the future and that could be put in a separate file by the generator, sure I'm for it. Although when I wrote the typedef generator I added `--##` at the top, I wonder if/why jarg removed that. But regardless of if it was removed in the generated type defs, ignoring builltin `---@meta` files is good anyway.

File: factorio-plugin/require.lua (hunk 1 above)
Comment (JanSharp) [reply factorio-plugin/require.lua:13]: Hm, maybe for factorio workspaces it should always be set to `/`. I've never tried `\` but I have a feeling not all zip libraries are going to accept that on all platforms. Something for me to keep in mind when rewriting the readme I suppose

File: plugin.lua [hunk 7]
```diff
@@ -5,54 +5,17 @@
...
-  if not package.path:find(new_path, 1, true) then
-    package.path = package.path..";"..new_path
-  end
+local fs = require("bee.filesystem")
+local workspace = require("workspace")
+local scope = require("workspace.scope")
+local plugin_path = fs.path(scope.getScope(workspace.rootUri):get('pluginPath'))
+local new_path = (plugin_path:parent_path() / "?.lua"):string()
+if not package.path:find(new_path, 1, true) then
+  package.path = package.path..";"..new_path
```
Comment (JanSharp) [inline plugin.lua:18]: I still don't get scopes in `sumneko.lua`. Like what you have here looks good but since I don't know what the point of scopes are (and you are using a scope here of course) I don't know if I can trust it

File: plugin.lua [hunk 8]
```diff
@@ -70,16 +33,16 @@ local on_event = require("factorio-plugin.on-event")
 ---@param text string @ The content of file
 ---@return nil|Diff[]
 function OnSetText(uri, text)
+
+  ---Not sure about this one if files are in the library it will not process
+  ---cross workspace requires seem to require the path in the library?
+  --TODO report that as sumneko bug? Needs more testing, maybe individual settings.
```
Comment (JanSharp) [inline plugin.lua]: raiguard and I have tested this yesterday and here's how it behaves:
`require`ing a file that is in a symlink within a library simply doesn't work, ever.
`require`ing a file using the `__modname__` prefix only works about 1/5 times on first load, but works 100% of the time when modifying the file with the `__modname__` require(s) in it. This appears to be some kind of race condition in `sumneko.lua` involving `require` resultion and plugin diffing.
(At least I'm assuming that's what you're talking about here, I'm not actually sure)

File: factorio-plugin/global.lua [hunk 9]
```diff
@@ -13,7 +13,7 @@ local function replace(uri, text, diffs)
       this_mod = "FallbackModName"
     else
       local workspace = require("workspace")
-      this_mod = (workspace.uri--[[@as string]] or workspace.getRootUri(uri)):match("[^/\\]+$")
+      this_mod = (workspace.rootUri or workspace.getRootUri(uri)):match("[^/\\]+$")
```
Comment (JanSharp) [inline factorio-plugin/global.lua]: `workspace.uri` is from `2.5.3` - it's intentional for backwards compatability. And btw this is actually broken, I fixed it in [this commit](https://github.com/JanSharp/FactorioSumnekoLuaPlugin/commit/d10b2c55832568447fda81a43c7eca1daa977251).

File: factorio-plugin/require.lua [hunk 10]
```diff
@@ -13,10 +13,7 @@ local function replace(_, text, diffs)
     local original_name = name
 
     ---Convert the mod name prefix if there is one
-    ---@param match string
-    name = name:gsub("^__(.-)__", function(match)
-      return match
-    end)
+    name = name:gsub("^__(.-)__", "%1", 1)
```
Comment (JanSharp) [inline factorio-plugin/require.lua]: lol, I wonder how that happened. This is either some technical debt or I was incredibly unaware of how gsub and patterns :D.
Oh also that 3rd parameter for gsub is optional and I prefer it being omitted if it's the default. Less code to read and comprehend

Comment (Nexela) [reply factorio-plugin/require.lua]: I totally missed that in this case it isn't needed since the ^ anchor is only going to limit it to 1 anyway

File: plugin.lua (hunk 8 above)
Comment (Nexela) [reply plugin.lua]: Somewhat
For the race condition you can see it with events if the file containing script.on_events is open sometimes after load/reload it won't update until you start typing.

`__modname__` requires seem to only look in the current scope,   if `__modname__` is your current scope and runtime.path includes your mods folder then magic happens
if `__modname__` is another workspace folder (and the path is set as above)  it will only read `__modname__` if it is included in workspace.library

What I have not tested yet was if  `__modname__` is the current scope and `__modname__` is in the library path does it get skipped with this check.

File: plugin.lua (hunk 5 above)
Comment (Nexela) [reply plugin.lua]: >However jarg has been talking about a custom ---@event tag which the plugin would have to see.

One of the things that my brain takes a second to process is `--##` Whats your thoughts on using something like
`---@plugin disable` and `---@plugin enable` to disable and enable blocks. It is more inline with with Sumneko styling and could be further expaned to `---@plugin disable: remote` to only disable it for a specific part of the plugin.

Comment (JanSharp) [reply plugin.lua]: The reason we started with `--##` was because that's the example (or was the example at least) in the `sumneko.lua` wiki for plugins. I've since kept it because it works and I didn't want to change it just to change it.
I've though about the enabling and disabling of the plugin or its modules for sections of a file but this isn't exactly easy to do with Lua patterns without sacraficing a notable amount of performance that scales with the amount of `---@plugin` tags in a file. It might not be super bad but I've never done it because I didn't see the need to. The cases where I saw a potential need to disable it I've added an "inline" `--##` to disable the plugin for that one case (like remote.add_interface). Plus my approach is that the plugin should be good enough to where you don't have the need to disable it which leads to fewer options for the user which makes it easier to use. And in my experience `--##` has only ever been used by myself to tell other people to test if the plugin is actually causing an issue, I've never heard or seen anything about it outside of that.
So TL;DR: I don't consider this to be an issue meaning increasing the complexity doesn't seem worth it

File: plugin.lua (hunk 7 above)
Comment (Nexela) [reply plugin.lua:18]: It can't be trusted. I think scope is only valid if the plugin is part of an opened workspace or library. At this point I should just PR the path in sumneko.

File: factorio-plugin/require.lua (hunk 10 above)
Comment (Nexela) [reply factorio-plugin/require.lua]: Ok after playing around with it some more. The modname identifier might have to be dropped in full.

Asumming `mods/ModA` and `mods/ModB` and runtime.path of `?.lua` and path strict true

If the workspace.library setting for ModA contains the root path `"/path/to/mods/"` then ModA and ModB get loaded twice in sumneko giving duplicate annotation errors however `require "__ModB__"` works

If the workspace.library setting for ModA contains `"/path/to/mods/ModB/"` then require doesn't work as it is looking for `/path/to/mods/ModB/ModB/file.lua`
Adding either `../?.lua` or `/path/to/mods/?.lua` to runtime.path does not work either.  This is possibly a bug or limitation in sumneko?

However if the require plugin does `__ModB__/file` -> `file` then it is correctly found.  This comes with the slight drawback that if both `__ModA__` and `__ModB__` have `file.lua` it needs to pointed in the right direction with an @as or @type on the require line.

Comment (JanSharp) [reply factorio-plugin/require.lua]: the strict path feature is broken with libraries - meaning it just doesn't find any of them when using require even without the plugin - and using a fully qualified path in `Lua.runtime.path` also simply doesn't work. This is my experience from a while ago and I could be remembering it wrong, but based on raiguard's experience with the strict path setting this still seems to be pretty accurte. I'd definitely call it a bug with `sumneko.lua` which has lead me to not use the strict path setting for the modding workspace

**Comment by JanSharp** [comment]
So now that I've decided I'm no longer going to use `require` in the plugin the only place using the "scope" is the `global` fallback which I currently wouldn't mind putting the current logic for getting the scope there (in a `pcall` most likely). Of course if there is an official way to get the scope at some point I'd switch to that.

With that in mind which parts of this PR would you say are still incomplete? I'd basically be waiting on that before I merge files into one.
(I've decided I'm not going to generate that file, I'm just going to merge them in source as well. Generating it is over complicating it way too much, the plugin really isn't that big to need multiple files).
But too be clear I'm not in a rush, just letting you know that I'm ready any time to merge the PR

**Comment by Nexela** [comment]
Disregarding the require stuff, good to go.
In its current state, I get 0 problems in many different workspace setups included, excluded, workspace code-workspace, .luarc settings, multiple different settings in a mult workspace
I don't even think a pcall is needed.

Language server runs once , and gathers everything into a collection
for each workspace (with a runtime.plugin defined) it re-loads the plugin into the environment for that workspace hence the multiple loads you see. This could be a sumneko bug, or it could not.

I.E if you have a 5 folders in a code.workspace file, and the code.workspace contains runtime.plugin it will get loaded into all 5 workspaces as each workspace is separate.  If workspace.runtime.plugin is blank but 3 workspaces have runtime.plugin in their settings.json it will only get loaded for those three.

**Comment by Nexela** [comment]
The best setup I have found was

Code-workspace requires the plugin
Code-workspace contains a a default set of libraries I want everything to use

If I want to cross-require in one workspace, I copy the code-workspace workspace settings into a settings.json for that mod
and add the the required library to it  i.e "relative/path/to/mod/b" from mod/a  But wait theres more
runtime.path needs to be set to either at the workspace level or folder level to include "?.lua" and "absolute/path/to/mods/?.lua"
settings from the closest scope are not merged with other scopes (yet, I put in a request to allow it)

**Comment by Nexela** [comment]
Re: Disregarding the require stuff.  I just hardcoded my path in and never got around to looking at it properly hoping for sumneko support or single file support

**Comment by Nexela** [comment]
~~Ok I think I broke something~~ Looks like I forgot to follow my own instructions and add the mods folder to runtime.path :P (actually don't know how it got removed)

**Comment by Nexela** [comment]
>runtime.path needs to be set to either at the workspace level or folder level to include "?.lua" and "absolute/path/to/mods/?.lua"

This was tested with runtime.pathstrict I don't know if its needed if that is disabled though

**Comment by Nexela** [comment]
Hmmm I think changing to `name = name:gsub("^__(.-)__[%./]", "")` will be the better way to go, as I can't get it to work without it, and adding the root folder to the library is just a no go. plus this will make for easier setup of not having to deal with runtime.path

**Comment by Nexela** [comment]
Bug found 2 the scope for global is wrong somehow

**Comment by Nexela** [comment]
Let me know your thoughts on the two recent commits, It should cover 85% of the use cases out there

**Comment by JanSharp** [comment]
I'm not sure about that global one. I mean I agree that if that previous code was broken it shouldn't exist, but what I had planned was to use the hack that I'm currently using to get the scope and use that for global. Current master is going that if you want to look again. Which is why I said pcall it, since it would break with internal changes in sumneko.lua (like it has at least 5 times by now).

The `__modname__` one is unfortuante but considering how gigantic of a pain it is to use the strict path setting it actually doesn't really make it all that much worse. Though I see you've dropped support for `\` :thinking: Which is guess is fine since using `\` in factorio is asking for trouble since it's all cross platform. Yea this is good.

**Comment by JanSharp** [comment]
Btw I'm not asking you to implement the hack for global, I actually plan on merging this PR and then making another PR myself just for that thing so we can have a discussion about it there

**Comment by Nexela** [comment]
The problem for both global and require is there are 2 ways to use a workspace:
 1. Each mod is opened as root in its own workspace or multi-workspace.
 2.  `mods_folder` is opened as root (the bleh way).

re: global
The correct way to get the path for global in either situation is the `workspace.getRootUri(uri)`
In a `mods_folder` as root workspace setup the getRootUri is `mods_folder`. If the `mods_folder` is named `mods` this works correctly due to the shortcut of trying to parse `mods` first before getRootUri.

re: require
Following that same path using `mods_folder` as root does allow the original `__modname__/a` -> `modname/a` to work with `path.strict=true/false` however the change to remove the mod identifier altogether https://github.com/JanSharp/FactorioSumnekoLuaPlugin/pull/3/commits/71643b11d68148cb3e74808e3a128d166576fda3 would only work with `path.strict = false` and I am not sure if I am a big fan of this option even though I don't use mods as root.

To summarize:
in option 1:  getRootUri is correct, and removing the modname from the require is correct.
in option 2: getRootUri is incorrect and removeing the modname from the require is incorrect.
 
Some options:
 1. Force the correct individual folder workspace setup.
 2. Have 2 versions of the plugin depending on your setup.
 3. Have a config variable in plugin.lua to allow the user set the mode between 1 or 2 from above.
 4. PR Sumneko to add support for calling an init function with config parameters.
 5. Do hacky failure prone things to get the path which I don't see working correctly ever.

I am all for `1 or 2 or 4`

**Comment by Nexela** [comment]
I think Sumneko pushed a fix that might handle the `requires` better ... but at the same time it broke plugin usage sooooooo

**Comment by Nexela** [comment]
Sumneko next release will support `runtime.path "../?.lua`

Current bug with sumeko is that it requires absolute path to plugin, hopefully it gets fixed

**Comment by Nexela** [comment]
I think I have a PR about ready for Sumneko that will add runtime.pluginArgs and even though you are possibly going to a single file adding the plugin path

**Comment by JanSharp** [comment]
Wait what, I wanted to know what that "resolve conficts" button does on github and once I resoved it it gave me a "commit merge" button, saying it would merge to your branch... and that's actually what it did?! I just made a commit on your repository without accepting a collaboration invite which means I'm pretty sure I don't have write access to your repo :thinking: very interesting

**Comment by JanSharp** [comment]
so I'm not entirely sure what to do. While this PR is ready and done I don't feel like it's actually in a good enough state to be released like this. `global` behaves differently and I wouldn't want to release an update for that and then like a week later release another one that changes it again ( #5 ). So I feel like the best bet would be to just wait until #5 is ready, right? It's not like current master is broken, it's just outdated

**Comment by Nexela** [comment]
This is a `meh` kind of call, I have thought of a way to automatically determine which mode to use `mods` or `folder` when the setting isn't passed, I haven't implemented it yet but it works in my head.

Have you come up with a decision on one plugin file or keeping the requires?

My camp is keep the separate requires. If @justarandomgeek can work with downloading it correctly. Of course this requires 3.5.next for requires to work correctly.  I am also of that camp that `user3dParty` should be used for both the typedefs and the plugin (which is currently broken in 3.5 but fixed for 3.5.next). This will handle loading the typedefs and the plugin without any additional settings needed other than a path to 3rd party and will allow Jarg to store the typedefs and plugins in workspace storage thingie from what I understand.

On the Sumneko side of things I am currently working on both adding a plugin sandbox as right now anything required inside a plugin is actually kept in package.loaded which doesn't get removed when a plugin is reset or another WS scope is loaded. And additionally adding plugin support for the color picker.
image
That WIP branch uses the color picker to update the table and I like it!  but the current events system would require plugin support to add different ways of representing the color picker  (another reason I am of team multiple files)

re: Sumneko Scopes
When a workspace is loaded it becomes the the first scope and workspace.rootURI and its children belongs to this scope. Additional workspaces add additional scopes. 3.5.1 will allow one scope to index into the other with `.../?.lua` in Lua.runtime.path`. I have not tested if this works for all scopes or just scopes that are part of the parent path.

Same for Sumneko settings I believe each scope reads first from `luarc.jsonc` then `workspace/.vscode/settings.lua` then `workspace.code-workspace settings` then from `user settings` then from `fallback settings (default)` The closest setting is not overwritten by settings up the line.

**Comment by justarandomgeek** [comment]
> Have you come up with a decision on one plugin file or keeping the requires?
> 
> My camp is keep the separate requires. If @justarandomgeek can work with downloading it correctly. 

I'm good with whatever style of packaging works best for everybody else, as long as there's an easy way for me to tell what version the current install is, whether there's a new version available, and if so where to get it from.

File: factorio-plugin/util.lua [hunk 11]
```diff
@@ -20,8 +20,7 @@
 local function gmatch_at_start_of_line(s, pattern, init)
   local first = true
   local unpack = table.unpack
-  ---@type fun(): string|integer, ...
-  local gmatch_iterator = s:gmatch("\n"..pattern, init)
+  local gmatch_iterator = s:gmatch("\n"..pattern) ---@type fun(): string|integer, ...
```
Comment (JanSharp) [inline factorio-plugin/util.lua:23]: What's the reason for removing `, init` here? In Lua `5.4` gmatch does take an init parameter and based on my testing from long ago it also exists in whatever versino of Lua sumneko.lua is using. It is interesting that nothing is actually using the init parameter in code right now, but if it's there it should still function as expected.
If the reason is that sumneko.lua's Lua actually doesn't have the init parameter for gmatch then let me know and I'll make sure to clean it up.
